pyscf_ipu/nanoDFT/nanoDFT.py

```diff
@@ -59,60 +72,78 @@ def nanoDFT_iteration(i, vals, opts, mol):
+    rho = jnp.sum(mult, axis=2) # (4, grid_size)=(4, 45624) for C6H6.
     E_xc, vrho, vgamma = b3lyp(rho, EPSILON_B3LYP)                                               # (gridsize,) (gridsize,) (gridsize,)
-    E_xc = jnp.sum(rho[0] * grid_weights * E_xc)                                                 # float=-27.968[Ha] for C6H6 at convergence.
-
+    E_xc = jax.lax.psum(jnp.sum(rho[0] * grid_weights * E_xc), axis_name="p")                                                 # float=-27.968[Ha] for C6H6 at convergence.
     rho = jnp.concatenate([vrho.reshape(1, -1)/2, 4*vgamma*rho[1:4]], axis=0) * grid_weights     # (4, grid_size)=(4, 45624)
-    V_xc = grid_AO[0].T @ jnp.sum(grid_AO * jnp.expand_dims(rho, axis=2), axis=0)                # (N, N)
+    grid_AO_T = grid_AO[0].T # (N, gsize)
+    rho = jnp.expand_dims(rho, axis=2) # (4, gsize, 1)
+    grid_AO_rho = grid_AO * rho # (4, gsize, N)
+    sum_grid_AO_rho = jnp.sum(grid_AO_rho, axis=0) # (gsize, N)
+    V_xc = grid_AO_T @ sum_grid_AO_rho # (N, N)
+    V_xc = jax.lax.psum(V_xc, axis_name="p") #(N, N)
     V_xc = V_xc + V_xc.T                                                                         # (N, N)
-
     return E_xc, V_xc                                                                            # (float) (N, N)
 
-def get_JK(density_matrix, ERI, opts, mol):
+def get_JK(density_matrix, ERI, dense_ERI, backend):
     """Computes the (N, N) matrices J and K. Density matrix is (N, N) and ERI is (N, N, N, N).  """
-    if opts.backend != "ipu":
+    N = density_matrix.shape[0]
+
+    if dense_ERI: 
         J = jnp.einsum('ijkl,ji->kl', ERI, density_matrix)                                       # (N, N)
         K = jnp.einsum('ijkl,jk->il', ERI, density_matrix)                                       # (N, N)
+        diff_JK = J - (K / 2 * HYB_B3LYP)
     else:
-        # Custom einsum which utilize ERI[ijkl]=ERI[ijlk]=ERI[jikl]=ERI[jilk]=ERI[lkij]=ERI[lkji]=ERI[lkij]=ERI[lkji]
-        J, K = ipu_einsum(ERI, density_matrix, mol, opts.threads, opts.multv)                    # (N, N) (N, N)
-    return J, K * HYB_B3LYP                                                                      # (N, N) (N, N)
+        from pyscf_ipu.nanoDFT.sparse_symmetric_ERI import sparse_symmetric_einsum
+        diff_JK = sparse_symmetric_einsum(ERI[0], ERI[1], density_matrix, backend)
+        
+    diff_JK = diff_JK.reshape(N, N)
+
+    return diff_JK
 
-def _nanoDFT(state, opts, mol):
+def _nanoDFT(state, ERI, grid_AO, grid_weights, opts, mol):
     # Utilize the IPUs MIMD parallism to compute the electron repulsion integrals (ERIs) in parallel.
-    if opts.backend == "ipu": state.ERI = electron_repulsion_integrals(state.input_floats, state.input_ints, mol, opts.threads_int, opts.intv)
-    else: pass # Compute on CPU.
+    #if opts.backend == "ipu": state.ERI = electron_repulsion_integrals(state.input_floats, state.input_ints, mol, opts.threads_int, opts.intv)
+    #else: pass # Compute on CPU.
```

Comment:
Maybe lose the comment at line 124 too? I.e. delete all of this.

Reply:
@mihaipgc  is working on rewriting `electron_repulsion_integrals(..)` to fit the layout in `sparse_symmetric_ERI(..)`; when we get that we'll revert back to these lines. 

After this, when @balancap allow tesselate to shard `jax.lax.gather_p` and `jax.lax.scater_add_p`, the DFT will never move ERI indices and ERI floats between tiles.